### PR TITLE
First draft

### PR DESCRIPTION
**Pull request by axfelix**

### CODE REVIEW

File: .bash.cfg
Comment (cretz) [inline .bash.cfg:1]: Fairly platform specific and may not be necessary

File: .gitignore
Comment (cretz) [inline .gitignore:1]: I think most of this gitignore wouldn't apply to this repo. I know this is a bog-standard one, but often in our repos we take a more measured approach of explicitly ignoring for the sake of our readers to see what may need to be ignored.

File: .vscode/settings.json
Comment (cretz) [inline .vscode/settings.json:1]: Usually IDE-specific settings files containing platform-specific things aren't the best in these repos

File: exercises/codec-server/README.md
Comment (cretz) [inline exercises/codec-server/README.md:1]: I wonder if there is value in linking to our sample repo or even the specific encryption sample

File: exercises/codec-server/solution/codec.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from typing import Iterable, List
+
+import cramjam
+from temporalio.api.common.v1 import Payload
+from temporalio.converter import PayloadCodec
+
+
+class EncryptionCodec(PayloadCodec):
```
Comment (cretz) [inline exercises/codec-server/solution/codec.py:8]: This is a class named `EncryptionCodec` that does compression (wish we could have encryption-only course and ignore compression)

File: exercises/codec-server/solution/worker.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+import temporalio.converter
+from temporalio import workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+from codec import EncryptionCodec
+
+
+@workflow.defn(name="Workflow")
```
Comment (cretz) [inline exercises/codec-server/solution/worker.py:12]: Not that you have to here (and we don't enough in samples), but we have been encouraging users to put workflows in their own module/file

File: exercises/custom-converter/README.md
Comment (cretz) [inline exercises/custom-converter/README.md:1]: Consider changing this dir to `exercises/custom-codec`. We have a concept of custom converters, but this isn't really that. _Technically_ it's a "custom" data converter because you have to create it to reference a custom codec, but usually by "custom converter" we mean custom payload converter and actually customizing object conversion (e.g. https://github.com/temporalio/samples-python/tree/main/custom_converter)

File: style.css
Comment (cretz) [inline style.css:1]: Is this file needed?

**Review by cretz — COMMENTED** [review]
Did quick review at high level, looks good overall, though I would like to see actual encryption though I know we're scared of choosing an encryption algorithm

File: .bash.cfg
Comment (axfelix) [reply .bash.cfg:1]: We just switched to using our own custom docker image across our courses, and this is in every language now for consistency. It does more in Java to set up aliases to avoid having to run maven commands, but it still gets used here.

File: .gitignore
Comment (axfelix) [reply .gitignore:1]: Fair observation!

File: .vscode/settings.json
Comment (axfelix) [reply .vscode/settings.json:1]: This is also relatively new along with the `.bash.cfg` — we're improving these as we go across all of our courses.

File: exercises/codec-server/solution/worker.py (hunk 2 above)
Comment (axfelix) [reply exercises/codec-server/solution/worker.py:12]: We do this consistently in every language but Python, for whatever reason. But we should get better about it in Python.

File: exercises/custom-converter/README.md
Comment (axfelix) [reply exercises/custom-converter/README.md:1]: Yup — this course evolved some as we shifted our focus to be on Codecs first and Payload Converters second (just as it evolved to only show compression and not encryption), and there are some lingering paths here.

File: style.css
Comment (axfelix) [reply style.css:1]: Yup, Gitpod CSS looks junky otherwise.

File: exercises/codec-server/README.md
Comment (axfelix) [reply exercises/codec-server/README.md:1]: We try not to link to *too* much outside material in our courses to keep them relatively linear (unlike the docs), but this would not be a bad place to make an exception, since the Encryption topic is very leading here. I'm considering it.

File: exercises/codec-server/solution/codec.py (hunk 1 above)
Comment (axfelix) [reply exercises/codec-server/solution/codec.py:8]: General point — it's less that we're scared of choosing an encryption algorithm (I actually make some recommendations from our Slack in https://github.com/temporalio/edu-appdatasec-go-content/blob/main/compression-and-encryption/data-encryption-methods.md) and more that proper key management is extremely out of scope for the course, and any encryption samples we could provide that do more than obfuscate bytes are much more problematically incomplete.

File: .vscode/settings.json
Comment (cretz) [reply .vscode/settings.json:1]: Can we remove pending improvement?

File: .bash.cfg
Comment (cretz) [reply .bash.cfg:1]: I noticed most of the issues are marked outdated but the files remain. Can remove the files pending the improvements?

File: exercises/codec-server/README.md
Comment (cretz) [reply exercises/codec-server/README.md:1]: Yeah, I've seen that, it can be tough for people to know where to learn more while they're learning. But for this one in particular, I think linking to more sample code from course code may have value.

File: exercises/codec-server/solution/codec.py (hunk 1 above)
Comment (cretz) [reply exercises/codec-server/solution/codec.py:8]: I don't think you need key management for an encryption sample, just something to show where they would apply encrypt/decrypt operations. I'd be ok if we even stubbed out the actual encryption to a "encryption" class whose "encrypt" and "decrypt" implementations are just base64. I think the primary goal should be to show "if you want encryption, here's how you plug it into Temporal" which I fear compression samples don't make clear.

File: exercises/codec-server/solution/worker.py (hunk 2 above)
Comment (cretz) [reply exercises/codec-server/solution/worker.py:12]: Yeah, different languages have different reasons for organizing code into files in their own ways, often idiomatic to the language or SDK. In this case, Python has a sandbox that works at the file level and therefore putting the workflow with the worker instead of separate is kinda an anti-pattern (one I know we follow in simple samples in the samples repo in the past, but don't really suggest in general).

File: exercises/custom-converter/README.md
Comment (cretz) [reply exercises/custom-converter/README.md:1]: :+1: Will wait to see fixed

File: style.css
Comment (cretz) [reply style.css:1]: Ug, yeah it's unfortunate that the code has to have large environment-specific things like this. Hopefully people leveraging this code can recognize the differences between files that are for them/learning and files that aren't.

File: .bash.cfg
Comment (axfelix) [reply .bash.cfg:1]: Well, no -- we use these deliberately to provide aliases and clean up the shell prompt in the Gitpod environment:

```
# Define some handy aliases for use in GitPod
alias workspace="cd ${GITPOD_REPO_ROOT}"
alias webui="gp preview $(gp url 8080)"

# Configure a shorter, but still useful, shell prompt
export PROMPT_DIRTRIM=2
export PS1="\[\033[01;34m\]\w\[\033[00m\]\$ "
```

File: exercises/codec-server/README.md
Comment (axfelix) [reply exercises/codec-server/README.md:1]: Ok, I'm putting this on the list of post-launch changes to these courses, because I think you're right, but a) I think this would belong in the course content rather than the exercise repo, and b) our encryption samples are inconsistent across languages and right now we're trying to harmonize these for release, so we'll plan to contribute more samples in our next update.

Comment (axfelix) [reply exercises/codec-server/README.md:1]: Merging this now in time for tomorrow's beta, but will address: https://github.com/temporalio/edu-appdatasec-python-content/pull/7

File: exercises/codec-server/solution/worker.py (hunk 2 above)
Comment (axfelix) [reply exercises/codec-server/solution/worker.py:12]: Merging now in time for tomorrow's beta, but will address: https://github.com/temporalio/edu-appdatasec-python-code/issues/3

File: exercises/codec-server/solution/codec.py (hunk 1 above)
Comment (cretz) [reply exercises/codec-server/solution/codec.py:8]: I think this is important to address. It can be unclear to a user to have a security course with no clear sample to implement security, only compression.

Comment (axfelix) [reply exercises/codec-server/solution/codec.py:8]: Yeah, good point. Done in https://github.com/temporalio/edu-appdatasec-python-code/pull/7.
